miniapps/dpg/util/complexstaticcond.hpp

```diff
@@ -0,0 +1,217 @@
+   Array<bool> IsTraceSpace;
+
+   // New set of "reduced" Finite Element Spaces
+   // (after static condensation)
+   Array<FiniteElementSpace *> tr_fes;
+
+   Array<int> dof_offsets;
+   Array<int> tdof_offsets;
+
+   Array<int> rdof_offsets;
+   Array<int> rtdof_offsets;
+
+   // Schur complement matrix
+   // S = A_ii - A_ib (A_bb)^{-1} A_bi.
+   BlockMatrix * S_r = nullptr;
+   BlockMatrix * S_i = nullptr;
+   BlockMatrix * S_e_r = nullptr;
+   BlockMatrix * S_e_i = nullptr;
+   ComplexOperator * S = nullptr;
+
+   BlockVector * y_r = nullptr;
+   BlockVector * y_i = nullptr;
+   Vector * y = nullptr;
+
+   Array<ComplexDenseMatrix * > lmat;
+   Array<Vector * > lvec;
+
+   Array<int> rdof_edof;      // Map from reduced dofs to exposed dofs
+   Array<int> ess_rtdof_list;
+
+   BlockMatrix * P = nullptr; // Block Prolongation
+   BlockMatrix * R = nullptr; // Block Restriction
+
+#ifdef MFEM_USE_MPI
+   BlockOperator * pS_r = nullptr;
+   BlockOperator * pS_e_r = nullptr;
+   BlockOperator * pS_i = nullptr;
+   BlockOperator * pS_e_i = nullptr;
+   // Block HypreParMatrix for Prolongation
+   BlockOperator * pP = nullptr;
+#endif
+
+   bool Parallel() const { return parallel; }
+
+
+   // tr_idx (trace dofs indices)
+   // int_idx (interior dof indices)
+   void GetReduceElementIndicesAndOffsets(int el, Array<int> & tr_idx,
+                                          Array<int> & int_idx,
+                                          Array<int> & offsets) const;
+
+   void GetReduceElementVDofs(int el, Array<int> & rdofs) const;
+   void GetElementVDofs(int el, Array<int> & vdofs) const;
+
+
+   //  S = A_ii - A_ib (A_bb)^{-1} A_bi.
+   //  y = y_i - A_ib (A_bb)^{-1} y_b
+   ComplexDenseMatrix * GetLocalShurComplement(int el, const Array<int> & tr_idx,
+                                               const Array<int> & int_idx,
+                                               const ComplexDenseMatrix & elmat,
+                                               const Vector & elvect_r,
+                                               const Vector & elvect_i,
+                                               Vector & rvect_r,
+                                               Vector & rvect_i);
+
+   void ComputeOffsets();
+
+   void BuildProlongation();
+#ifdef MFEM_USE_MPI
+   void BuildParallelProlongation();
+#endif
+
+   // ess_tdof list for each space
+   Array<Array<int> *> ess_tdofs;
+   void FillEssTdofLists(const Array<int> & ess_tdof_list);
+
+   void ConformingAssemble(int skip_zeros);
+
+   /** Restrict a marker Array on the true FE spaces dofs to a marker Array on
+    the reduced/trace true FE spaces dofs. */
+   void ConvertMarkerToReducedTrueDofs(Array<int> & tdof_marker,
+                                       Array<int> & rtdof_marker);
+public:
+
+   ComplexBlockStaticCondensation(Array<FiniteElementSpace *> & fes_);
+
+   ~ComplexBlockStaticCondensation();
+
+   void SetSpaces(Array<FiniteElementSpace*> & fes_);
+
+   void Init();
```

Comment:
Document when this should be called or remove it from the `public` interface if it is only used internally.